### PR TITLE
boot: Change GUID handling

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

**Comment by bluca** [comment]
if you rebase now, build should be fixed

File: src/boot/efi/boot.c [hunk 1]
```diff
@@ -1576,6 +1576,7 @@ static void config_load_defaults(Config *config, EFI_FILE *root_dir) {
         _cleanup_free_ char *content = NULL;
         UINTN value;
         EFI_STATUS err;
+        EFI_GUID guid = LOADER_GUID;
```
Comment (zx2c4) [inline src/boot/efi/boot.c:1579]: Why not always use the MAKE_GUID macro? The compiler should do the right thing, yea? And it's not like guid is reassigned anywhere that would make using a variable good for the logical flow

File: src/boot/efi/boot.c [hunk 2]
```diff
@@ -2560,27 +2562,28 @@ static void export_variables(
 
         _cleanup_free_ char16_t *infostr = NULL, *typestr = NULL;
         char16_t uuid[37];
+        EFI_GUID guid = LOADER_GUID;
```
Comment (zx2c4) [inline src/boot/efi/boot.c:2565]: Ditto here.

File: src/boot/efi/devicetree.c [hunk 3]
```diff
@@ -9,8 +9,9 @@
 #define FDT_V1_SIZE (7*4)
 
 static void *get_dtb_table(void) {
+        EFI_GUID guid = EFI_DTB_TABLE_GUID;
```
Comment (zx2c4) [inline src/boot/efi/devicetree.c:12]: Is the memcmp actually more efficient from declaring a variable?  I kind of doubt it.

Also, opportunity to switch this function to size_t.

File: src/boot/efi/random-seed.c [hunk 4]
```diff
@@ -12,8 +12,6 @@
 #define RANDOM_MAX_SIZE_MIN (32U)
 #define RANDOM_MAX_SIZE_MAX (32U*1024U)
 
-#define EFI_RNG_GUID &(const EFI_GUID) EFI_RNG_PROTOCOL_GUID
```
Comment (zx2c4) [inline src/boot/efi/random-seed.c:15]: The superfluousness of this annoyed me before, so thanks for fixing it.

File: src/boot/efi/stub.c [hunk 5]
```diff
@@ -87,47 +87,48 @@ static void export_variables(EFI_LOADED_IMAGE_PROTOCOL *loaded_image) {
                 0;
 
         char16_t uuid[37];
+        EFI_GUID guid = LOADER_GUID;
```
Comment (zx2c4) [inline src/boot/efi/stub.c:90]: Doesn't need to be a variable.

File: src/boot/efi/xbootldr.c [hunk 6]
```diff
@@ -133,7 +133,7 @@ static EFI_STATUS try_gpt(
                 EFI_PARTITION_ENTRY *entry =
                                 (EFI_PARTITION_ENTRY *) ((uint8_t *) entries + gpt.gpt_header.SizeOfPartitionEntry * i);
 
-                if (memcmp(&entry->PartitionTypeGUID, XBOOTLDR_GUID, sizeof(entry->PartitionTypeGUID)) != 0)
+                if (memcmp(&entry->PartitionTypeGUID, &MAKE_GUID(XBOOTLDR), sizeof(entry->PartitionTypeGUID)) != 0)
```
Comment (zx2c4) [inline src/boot/efi/xbootldr.c:136]: Maybe you can fold in @poettering's https://github.com/systemd/systemd/pull/25350 to this commit, except make it lass by value, and then you can clean up all these memcmps to do the right thing too.

Then the rule is: GUIDs always pass by reference, with the one exception being EFI function pointer prototypes that we can't change.

File: src/boot/efi/util.h [hunk 7]
```diff
@@ -129,14 +129,16 @@ static inline void unload_imagep(EFI_HANDLE *image) {
                 (void) BS->UnloadImage(*image);
 }
 
+#define MAKE_GUID(GUID) ((EFI_GUID) GUID##_GUID)
```
Comment (poettering) [inline src/boot/efi/util.h:132]: so if we decide to pass around guids as in userspace, then that#s cetrainly something we can discuss. But if so, then let's try to make the macro more careful, i.e. let's keep the const in to give the compiler the hint that the literal data actually *is* constant, but then cast the const away to work around UEFI's stupid api definitoins.

i.e. maybe like so:

```c
#define MAKE_CONST_GUID(g) ((const EFI_GUID) g##_GUID)
#define MAKE_GUID(g) (*(EFI_GIUD*) &MAKE_CONST_GUID(g))
```

untested, but I think this should work.

**Comment by medhefgo** [comment]
I think I'll abandon this approach. Mixing copy-by-value and copy-by-ref is annoying. Instead a future PR will change all cases this to this:
```c
#define MAKE_GUID_PTR(name) ((EFI_GUID *) &(const EFI_GUID) name##_GUID)
…
func(MAKE_GUID_PTR(EFI_SOME_PROTO_NAME));
```

**Pull request closed without merging**